core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala

```diff
@@ -337,38 +332,66 @@ trait AttributeComponent {
-                on ta.transaction_id = $transactionId and (a.namespace,a.name,a.owner_id,ifnull(a.list_index,-2))=(ta.namespace,ta.name,ta.owner_id,ifnull(ta.list_index,-2))
-                where ta.owner_id is null and a.owner_id in """, ownerIdTail(ownerIds)).as[Int]
-      def updateInMasterAction(ownerIds: Seq[OWNER_ID], transactionId: String) =
+
+      // updateInMasterAction: updates any row in *_ATTRIBUTE that also exists in *_ATTRIBUTE_SCRATCH
+      def updateInMasterAction(transactionId: String) =
         sql"""update #${baseTableRow.tableName} a
                 join #${baseTableRow.tableName}_SCRATCH ta
                 on (a.namespace,a.name,a.owner_id,ifnull(a.list_index,-2))=(ta.namespace,ta.name,ta.owner_id,ifnull(ta.list_index,-2)) and ta.transaction_id = $transactionId
                 set a.value_string=ta.value_string, a.value_number=ta.value_number, a.value_boolean=ta.value_boolean, a.value_entity_ref=ta.value_entity_ref, a.list_length=ta.list_length""".as[Int]
-      def insertIntoMasterAction(ownerIds: Seq[OWNER_ID], transactionId: String) =
-        concatSqlActions(sql"""insert into #${baseTableRow.tableName}(namespace,name,value_string,value_number,value_boolean,value_entity_ref,list_index,owner_id,list_length)
-                select ta.namespace,ta.name,ta.value_string,ta.value_number,ta.value_boolean,ta.value_entity_ref,ta.list_index,ta.owner_id,ta.list_length
-                from #${baseTableRow.tableName}_SCRATCH ta
-                left join #${baseTableRow.tableName} a
-                on (a.namespace,a.name,a.owner_id,a.list_index)<=>(ta.namespace,ta.name,ta.owner_id,ta.list_index) and a.owner_id in """, ownerIdTail(ownerIds),
-                sql""" where ta.transaction_id = $transactionId and a.owner_id is null""").as[Int]
 
       def clearAttributeScratchTableAction(transactionId: String) = {
         sqlu"""delete from #${baseTableRow.tableName}_SCRATCH where transaction_id = $transactionId"""
       }
 
-      def upsertAction(ownerIds: Seq[OWNER_ID], insertFunction: String => ReadWriteAction[Unit]) = {
+      def updateAction(insertIntoScratchFunction: String => ReadWriteAction[Unit]) = {
         val transactionId = UUID.randomUUID().toString
-        insertFunction(transactionId) andThen
-          deleteFromMasterAction(ownerIds, transactionId) andThen
-          updateInMasterAction(ownerIds, transactionId) andThen
-          insertIntoMasterAction(ownerIds, transactionId) andFinally
+        insertIntoScratchFunction(transactionId) andThen
+          updateInMasterAction(transactionId) andThen
           clearAttributeScratchTableAction(transactionId)
       }
     }
 
+    // enables set operations on attribute primary key only
+
+    trait AttributeRecordPrimaryKey[ID_TYPE] {    // can't call it OWNER_ID because that is already in scope
+      val ownerId: ID_TYPE
+      val namespace: String
+      val name: String
+      val listIndex: Option[Int]
```

Comment:
I'm using the primary key which does not include listLength.  Whether it should, I don't know.